README.md
This is a module for Jelix, providing a plugin for jAuth allowing to authenticate
with an ldap server, and register them in the app database, using a dao.

This module is for Jelix 1.6.x and higher. 


Installation
============

Configuration with Jelix 1.7
----------------------------

You should use Composer to install the module. Run this commands in a shell:
                                               
```
composer require "jelix/ldapdao-module"
```

Launch the configurator for your application to enable the module

```bash
php yourapp/dev.php module:configure ldapdao
```


Configuration with Jelix 1.6
-----------------------------

Copy the `ldapdao` directory into the modules/ directory of your application.

Next you must say to Jelix that you want to use the module. Declare
it into the `var/config/mainconfig.ini.php` file.

In the `[modules]` section, add:

```ini
ldapdao.access=1
```

Following modules are required: jacl2, jauth, jauthdb. In this same section 
verify that they are activated:

```ini
jacl2.access=1
jauth.access=2
jauthdb.access=1
```

If you are using the jCommunity module, you must not activate jauth and jauthdb, 
so keep `jauth.access=0` and `jauthdb.access=0`.


Launch the installer
--------------------

In the command line, launch:

```
php yourapp/install/installer.php
```

Configuration
=============

This module provides two things:

1. a plugin, ```ldapdao```, for jAuth
2. a configuration file for the ```auth``` plugin for jCoordinator.

The ```ldapdao``` plugin replaces the `db` or `ldap` plugin for jAuth. The 
installer of the module deactivates some jAcl2 rights, and copy an example 
of the configuration file `authldap.coord.ini.php` into the configuration directory  
(`var/config` in Jelix 1.6, `app/system` in Jelix 1.7).

You should edit the new file `authldap.coord.ini.php`. Many properties
should be changed to match your ldap structure.

Second, you should indicate this new configuration file into the mainconfig.ini.php file,
in the `coordplugins` section:

```
[coordplugins]
auth="authldap.coord.ini.php"
```

General configuration properties
---------------------------------

First you should set the `dao`, `profile`, `ldapprofile` and `form` properties
in the `ldapdao` section of `authldap.coord.ini.php`, to indicate the dao 
(for the table), the form (for the administration module) and profiles to access 
to the database and the ldap.

Here is an example:

```
[ldapdao]

; name of the dao to get user data. It may differ depending
; to the application
dao = "jauthdb~jelixuser"

; name of the form for the jauthdb_admin module. It may differ depending
; to the application
form = "jauthdb_admin~jelixuser"

; profile to use for jDb 
profile = "myldapdao"

; profile to use for ldap
ldapprofile = "ldapdao"

```

For profiles, you should set connections parameters into the 
`var/config/profiles.ini.php` file.

Example of a profile for the dao:

```
[jdb:myldapdao]
driver="mysqli"
host= "localhost"
database="userdb"
user= "admin"
password="jelix"
persistent= on
force_encoding = on
```

Here the profile is named `myldapdao` so you should set `profile=myldapdao` into
`authldap.coord.ini.php`.

Connection configuration
------------------------

You should create an ldap profile, with the name indicated into the `ldapprofile`
parameter. The ldap profile should be into the `var/config/profiles.ini.php` file.

For example, if the profile is named `myldap` (like in the following example), 
so you should set `ldapprofile=myldap` into `authldap.coord.ini.php`.

Example of a profile for the ldap connection:

```
[ldap:ldapdao]
hostname=localhost
port=389
tlsMode=starttls  ; empty, "starttls" or "ldaps" (ldaps by default if port 636)
adminUserDn="cn=admin,ou=admins,dc=acme"
adminPassword="Sup3rP4ssw0rd"
searchUserBaseDN="dc=XY,dc=fr"
searchUserFilter="(&(objectClass=posixAccount)(uid=%%LOGIN%%))"
bindUserDN="uid=%?%,ou=users,dc=XY,dc=fr"
searchAttributes="uid:login,givenName:firstname,sn:lastname,mail:email"
searchGroupFilter="(&(objectClass=posixGroup)(cn=XYZ*)(memberUid=%%LOGIN%%))"
searchGroupProperty="cn"
searchGroupBaseDN=""
```

First, you should set `hostname`, `port`, which are the name and the port of
the ldap server.

You must also indicate how the connexion should be made. You should indicate 
if you are using the ldap protocol without encryption (`tlsmode=`), the ldap 
protocol with STARTTLS, (`tlsMode=starttls`) or if you are using the ldaps 
protocol (`tlsMode=ldaps`).

If you use the standard port 389, you have the choice between no encryption (empty value)
or `starttls`, for the `tlsMode` setting.

If you are using the port 636, it automatically uses the ldaps protocol. In this
case, setting `tlsMode` to `ldaps` is optional.

Since ldaps is deprecated in OpenLdap, the best solution is to use the port 389
with `tlsMode=starttls`.



Configuration properties for the admin
--------------------------------------

The plugin needs to query the directory with a user having some rights allowing 
to search a user, to get his attributes, his groups etc... He's called the "admin" 
in the plugin.

You must indicate in `adminUserDn` and `adminPassword`, the DN (Distinguished Name) 
and the password of this user. 


Configuration properties for user data
--------------------------------------

To verify password, or to register the user into the Jelix application the
first time he authenticate himself, the plugin needs some data about the user.

You should indicate to it which ldap attributes it can retrieve, and which
database fields that will receive the ldap attributes values.

You indicate such informations into the `searchAttributes` property. It is a 
pair of names, `<ldap attribute>:<table field>`, separated by a comma.

In this example, `searchAttributes="uid:login,givenName,sn:lastname,mail:email,dn:"`:

- the value of the `uid` ldap attribute will be stored into the `login` field 
- the value of the `sn` ldap attribute will be stored into the `lastname` field
- the value of the `givenName` ldap attribute will be stored into a field that
  have the same name, as there is no field name nor `:`.
- there will not be mapping for the `dn` property. There is a `:` without field name.
  It will be readed from ldap, and can be used into the `bindUserDN` DN template.
  (see below).

The possible list of possible fields is indicated into the dao file, whose name
is indicated into the `dao` configuration property.


Configuration properties for authentication
-------------------------------------------

Starting from 2.0.0, the login process has changed, to take care of various
ldap structure and server configuration.

Before to try to authenticate the user against the ldap, the plugin retrieves
user properties. It uses two configuration parameters : `searchUserFilter`
and `searchAttributes`.

The `searchUserFilter` should contain the ldap query, and a `%%LOGIN%%` placeholder
that will be replaced by the login given by the user.

Example: `searchUserFilter="(&(objectClass=posixAccount)(uid=%%LOGIN%%))"`

You may also indicate the base DN for the search, into `searchUserBaseDN`. Example:
`searchUserBaseDN="ou=ADAM users,o=Microsoft,c=US"`.

Note that you can indicate several search filters, if you have
complex ldap structure. Use `[]` to indicate an item list:

```
searchUserFilter[]="(&(objectClass=posixAccount)(uid=%%LOGIN%%))"
searchUserFilter[]="(&(objectClass=posixAccount)(cn=%%LOGIN%%))"
```

To verify the password, the plugin needs the DN (Distinguished Name) corresponding 
to the user. It builds the DN from a "template" indicated into the `bindUserDN`
property, and from various data. These data can be the given login or one of
the ldap attributes of the user.

- Building the DN from the login given by the user: bindUserDN should contain
  a DN, with a `%%LOGIN%%` placeholder that will be replaced by the login.

  Example: `bindUserDN="uid=%%LOGIN%%,ou=users,dc=XY,dc=fr"`. If the user
  give `john.smith` as a login, the authentication will be made with the DN
  `bindUserDN="uid=john.smith,ou=users,dc=XY,dc=fr"`.

  For some LDAP, the DN could be a simple string, for example an email. 
  You could then set `bindUserDN="%%LOGIN%%@company.local"`. Or even 
  `bindUserDN="%%LOGIN%%"` if the user can type the full value of
  the DN or an email or else.. (Probably it's not recommended to allow
  a user to type himself its full DN, it can be a security issue)

- Building the DN from one of the ldap attributes of the user. 
  In this case, the plugin will first query the ldap directory with the 
  `searchUserFilter` filter, to retrieve the user's ldap attributes.
  Then, in bindUserDN, you can indicate a DN where some values will be replaced
  by some attributes values, or you can indicate a single attribute name,
  corresponding to an attribute that contain the full DN of the user.
  
  For the first case, bindUserDn should contain a DN, with some `%?%` placeholders
  that will be replaced by corresponding attributes value. Example:
  `bindUserDN="uid=%?%,ou=users,dc=XY,dc=fr"`. Here it replaces the `%?%` by the
  value of the `uid` attribute readed from the user's attributes.
  The attribute name should be present into the `searchAttributes`
  configuration property, even with no field mapping. Ex: `...,uid,...`. See above.
  
  For the second case, just indicate the attribute name, prefixed with a `$`.
  Example: `bindUserDN="$dn"`. Here it takes the `dn` attribute readed from 
  the search, and use its full value as the DN to login against the ldap server.
  It is useful for some LDAP server like sometimes Active Directory that need a 
  full DN specific for each user.
  The attribute name should be present into the `searchAttributes`
  configuration property, even with no field mapping. Ex: `...,dn:,...`. See above.

Note that you can indicate several dn templates, if you have
complex ldap structure. Use `[]` to indicate an item list:

```
bindUserDN[]="uid=%?%,ou=users,dc=XY,dc=fr"
bindUserDN[]="cn=%?%,ou=users,dc=XY,dc=fr"
```

Configuration properties for user rights
----------------------------------------

If you have configured groups rights into your application, and if these
groups match your ldap groups, you can indicate to the plugin to automatically
put the user into the application groups, according to the user ldap groups.

You should then indicate into `searchGroupFilter` the ldap query that will
retrieve the groups of the user.

Example: `searchGroupFilter="(&(objectClass=posixGroup)(member=%%USERDN%%))"`

`%%USERDN%%` is replaced by the user dn.`%%LOGIN%%` is replaced by the login.  
You can also use any ldap attributes you indicate into `searchAttributes`, 
between `%%`. Example: `searchGroupFilter="(&(objectClass=posixGroup)(member=%%givenName%%))"`

Warning : setting `searchGroupFilter` will remove the user from any other
application groups that don't match the ldap group, except default groups
of user into the application, when `searchGroupKeepUserInDefaultGroups` is set 
to `on`. If you don't want a groups synchronization, leave `searchGroupFilter` empty.

With `searchGroupProperty`, you must indicate the ldap attribute that
contains the group name. Ex: `searchGroupProperty="cn"`.

You may also indicate the base DN for the search, into `searchGroupBaseDN`. Example:
`searchGroupBaseDN="ou=Groups,dc=Acme,dc=pt"`.



CHANGELOG.md
Changes
=======

Version 2.2.5 (2023-10-09)
-------------------------

- Adds the `getDao()` method on the authentication plugin for applications that are using Jelix 1.8.3.


Version 2.2.4 (2023-04-11)
--------------------------

Installer: set automatically the jAuth driver value, even in "noconfigfile" mode.

Version 2.2.3 (2023-02-17)
--------------------------

- no interactive configuration, it is useless

Version 2.2.2 (2022-06-28)
--------------------------

- Fix the ldapdao installer for Jelix 1.7+

Version 2.2.1 (2021-09-16)
--------------------------

- Fix an error during the installation


Version 2.2.0 (2020-11-09)
--------------------------

- Implements jIAuthDriver2 interface available in Jelix 1.6.21+


Version 2.1.1 (2020-04-16)
---------------------------

- Fix password verification: close ldap connection in case of errors
- Fix error message when the loading of user attributes fails

Version 2.1.0 (2020-03-24)
---------------------------

Compatibility with Jelix 1.7.x


Version 2.0.10 (2020-03-24)
----------------------------

remove the dependency to jauth in module.xml, as it is a conflict when
jcommunity is installed.


Version 2.0.9 (2020-02-29)
--------------------------

Add more messages into logs when something wrong happens, to ease the debugging 
of bad connections / authentication / search.

Version 2.0.8 (2019-11-18)
--------------------------

- Improve support of ldaps and STARTTLS, by using ldap uri and ldap_start_tls()


Version 2.0.7 (2019-04-18)
--------------------------

- Fix SQL error during the installation


Version 2.0.6 (2019-01-28)
--------------------------

- Security issue: empty password may be accepted by LDAP servers, as they 
  implement the Unauthenticated Authentication Mecanism, allowing empty password.
  We don't want such feature, so empty password are now refused. 
- Fix user group search: escape parenthesis in values inserted into the 
  searchGroupFilter name.
- Fix user group search: add the user in default groups of the application,
  in addition to groups corresponding to ldap groups. New configuration 
  parameter searchGroupKeepUserInDefaultGroups to disable this feature.

Version 2.0.5 (2018-11-20)
--------------------------

- fix error during the installation

Version 2.0.4 (2018-07-27)
--------------------------

- fix some unit tests

Version 2.0.3 (2018-07-03)
--------------------------

- Fix: allow the admin user to change his password
- Fix: in some case, the user was recognized only at the first login

Version 2.0.2 (2018-06-08)
--------------------------

- Fix support of jCommunity

Version 2.0.1 (2018-01-29)
--------------------------

- Fix: attributes search were made anonymously, which is not allowed in some 
  LDAP server
- Fix support of attributes declaration having no mapping like `foo:`
- Improved the configuration manual
- Fix some unit tests


Version 2.0.0 (2017-04-05)
--------------------------

The login process has changed, to take care of various ldap structure and 
server configuration.

- support of multiple search filters for users
- support of multiple dn templates to connect users
- move ldap connection parameters (hostname, port, admin login/password)
  to profiles.ini.php
- Jelix admin login is configurable
- synchronize all ldap groups into jAcl2 rights, if configured

Version 1.1.1 (2017-02-08)
--------------------------

- Fix mispelling variable names


Version 1.1.0 (2015-06-03)
--------------------------

Initial public release.


tests/unittests/bootstrap.php
<?php
require_once(__DIR__.'/../testapp/application.init.php');
require_once(LIB_PATH.'jelix-tests/classes/junittestcase.class.php');
require_once(LIB_PATH.'jelix-tests/classes/junittestcasedb.class.php');

define('TESTAPP_URL', 'http://ldapdao.local/');
define('TESTAPP_URL_HOST', 'ldapdao.local');
define('TESTAPP_URL_HOST_PORT', 'ldapdao.local');
define('TESTAPP_HOST', 'ldapdao.local');
define('TESTAPP_PORT', '');

define('TESTAPP_LDAP_HOST', 'ldap.jelix');


jApp::setEnv('jelixtests');
if (file_exists(jApp::tempPath())) {
    jAppManager::clearTemp(jApp::tempPath());
} else {
    jFile::createDir(jApp::tempPath(), intval("775",8));
}


ldapdao/plugins/auth/ldapdao/ldapdao.auth.php
<?php
/**
 */

/**
 * LDAP authentification driver for authentification information stored in LDAP server
 * and manage user locally with a dao
 *
 *
 *
 * @package    jelix
 * @subpackage auth_driver
 *
 * @internal see https://tools.ietf.org/html/rfc4510
 */
class ldapdaoAuthDriver extends jAuthDriverBase implements jIAuthDriver2
{

    /**
     * default user attributes list
     * @var array
     */
    protected $_default_attributes = array(
        "cn"=>"lastname",
        "name"=>"firstname"
    );

    protected $uriConnect = '';

    public function __construct($params)
    {
        if (!extension_loaded('ldap')) {
            throw new jException('ldapdao~errors.extension.unloaded');
        }

        parent::__construct($params);

        if (!isset($this->_params['ldapprofile']) || $this->_params['ldapprofile'] == '') {
            throw new jException('ldapdao~errors.ldap.profile.missing');
        }

        $profile = jProfiles::get('ldap', $this->_params['ldapprofile']);
        $this->_params = array_merge($this->_params, $profile);

        // default ldap parameters
        $_default_params = array(
            'hostname'      =>  'localhost',
            'tlsMode'       => '',
            'port'          =>  389,
            'adminUserDn'      =>  null,
            'adminPassword'      =>  null,
            'protocolVersion'   =>  3,
            'searchUserBaseDN' => '',
            'searchGroupFilter' => '',
            'searchGroupKeepUserInDefaultGroups' => true,
            'searchGroupProperty' => '',
            'searchGroupBaseDN' => ''
        );

        // iterate each default parameter and apply it to actual params if missing in $params.
        foreach ($_default_params as $name => $value) {
            if (!isset($this->_params[$name]) || $this->_params[$name] == '') {
                $this->_params[$name] = $value;
            }
        }

        if ($this->_params['searchUserBaseDN'] == '') {
            throw new jException('ldapdao~errors.search.base.missing');
        }

        if (!isset($this->_params['searchAttributes']) || $this->_params['searchAttributes'] == '') {
            $this->_params['searchAttributes'] = $this->_default_attributes;
        } else {
            $attrs = explode(",", $this->_params['searchAttributes']);
            $this->_params['searchAttributes'] = array();
            foreach ($attrs as $attr) {
                if (strpos($attr, ':') === false) {
                    $attr = trim($attr);
                    $this->_params['searchAttributes'][$attr] = $attr;
                } else {
                    $attr = explode(':', $attr);
                    $this->_params['searchAttributes'][trim($attr[0])] = trim($attr[1]);
                }
            }
        }

        if (!isset($this->_params['searchUserFilter']) || $this->_params['searchUserFilter'] == '') {
            throw new jException('ldapdao~errors.searchUserFilter.missing');
        }
        if (!is_array($this->_params['searchUserFilter'])) {
            $this->_params['searchUserFilter'] = array($this->_params['searchUserFilter']);
        }

        if (!isset($this->_params['bindUserDN']) || $this->_params['bindUserDN'] == '') {
            throw new jException('ldapdao~errors.bindUserDN.missing');
        }
        if (!is_array($this->_params['bindUserDN'])) {
            $this->_params['bindUserDN'] = array($this->_params['bindUserDN']);
        }

        $uri = $this->_params['hostname'];

        if (preg_match('!^ldap(s?)://!', $uri, $m)) { // old way to specify ldaps protocol
            $predefinedPort = '';
            if (preg_match('!:(\d+)/?!', $uri, $mp)) {
                $predefinedPort = $mp[1];
            }
            if (isset($m[1]) && $m[1] == 's') {
                $this->_params['tlsMode'] = 'ldaps';
            }
            elseif ($this->_params['tlsMode'] == 'ldaps') {
                $this->_params['tlsMode'] = 'starttls';
            }
            if ($predefinedPort == '') {
                $uri .= ':'.$this->_params['port'];
            }
            else {
                $this->_params['port'] = $predefinedPort;
            }
            $this->uriConnect = $uri;
        }
        else {
            $uri .= ':'.$this->_params['port'];
            if ($this->_params['tlsMode'] == 'ldaps' || $this->_params['port'] == 636 ) {
                $this->uriConnect = 'ldaps://'.$uri;
                $this->_params['tlsMode'] = 'ldaps';
            }
            else {
                $this->uriConnect = 'ldap://'.$uri;
            }
        }
    }

    /**
     * @return jDaoFactoryBase
     */
    public function getDao()
    {
        return jDao::get($this->_params['dao'], $this->_params['profile']);
    }

    public function saveNewUser($user)
    {
        throw new jException("ldapdao~errors.unsupported.user.creation");
    }

    public function removeUser($login)
    {
        $this->getDao()->deleteByLogin($login);
        return true;
    }

    public function updateUser($user)
    {
        if (!is_object($user)) {
            throw new jException('ldapdao~errors.object.user.unknown');
        }

        if ($user->login == '') {
            throw new jException('ldapdao~errors.user.login.unset');
        }

        $this->getDao()->update($user);
        return true;
    }

    public function getUser($login)
    {
        $dao = $this->getDao();
        $user = $dao->getByLogin($login);
        if ($user) {
            return $user;
        }

        $connect = $this->_bindLdapAdminUser();
        if ($connect === false) {
            return false;
        }
        $user = $this->createUserObject($login, '');
        $found = $this->searchLdapUserAttributes($connect, $login, $user);
        ldap_close($connect);
        if (!$found) {
            return false;
        }
        $dao->insert($user);
        return $user;
    }

    public function createUserObject($login, $password)
    {
        $user = jDao::createRecord($this->_params['dao'], $this->_params['profile']);
        $user->login = $login;
        // should not be empty because of a jauth listener that prevent
        // user not having password to login.
        $user->password = '!!ldapdao password!!';
        return $user;
    }

    public function getUserList($pattern)
    {
        $dao = $this->getDao();
        if ($pattern == '%' || $pattern == '') {
            return $dao->findAll();
        } else {
            return $dao->findByLogin($pattern);
        }
    }

    public function canChangePassword($login)
    {
        if ($login == $this->_params['jelixAdminLogin']) {
            return true;
        }
        return false;
    }

    public function changePassword($login, $newpassword)
    {
        if ($login == $this->_params['jelixAdminLogin']) {
            return $this->getDao()->updatePassword($login, $this->cryptPassword($newpassword));
        }

        throw new jException('ldapdao~errors.unsupported.password.change');
    }

    public function verifyPassword($login, $password)
    {
        $dao = $this->getDao();

        if ($login == $this->_params['jelixAdminLogin']) {
            $user = $dao->getByLogin($login);
            return $this->checkAdminLogin($user, $dao, $password);
        }

        if (trim($password) == '' || trim($login) == '') {
            // we don't want Unauthenticated Authentication
            // and Anonymous Authentication
            // https://tools.ietf.org/html/rfc4513#section-5.1
            return false;
        }

        $connectAdmin = $this->_bindLdapAdminUser();
        if (!$connectAdmin) {
            return false;
        }

        // see if the user exists into the ldap directory
        $user = $this->createUserObject($login, '');
        $userLdapAttributes = $this->searchLdapUserAttributes($connectAdmin, $login, $user);
        if ($userLdapAttributes === false) {
            ldap_close($connectAdmin);
            return false;
        }

        $connect = $this->_getLinkId();
        if (!$connect) {
            ldap_close($connectAdmin);
            return false;
        }
        // authenticate user. let's try with all configured DN
        $userDn = $this->bindUser($connect, $userLdapAttributes, $user->login, $password);
        ldap_close($connect);

        if ($userDn === false) {
            jLog::log('ldapdao error: cannot authenticate to ldap with given bindUserDN for the login "' . $login. '". Wrong DN or password', 'auth');
            foreach ($this->bindUserDnTries as $erroMessage) {
                jLog::log($erroMessage, 'auth');
            }
            ldap_close($connectAdmin);
            return false;
        }

        // check if he is in our database
        $userDb = $dao->getByLogin($user->login);
        if (!$userDb) {
            if (jApp::isModuleEnabled('jcommunity')) {
                $user->status = 1; // STATUS_VALID
            }
            $dao->insert($user);
            $eventResp = jEvent::notify('AuthNewUser', array('user' => $user));
            $allResponses = array();
            if ($eventResp->inResponse('doUpdate', true, $allResponses)) {
                $dao->update($user);
            }
        }
        else {
            $user = $userDb;
        }

        // retrieve the user group (if relevant)
        $userGroups = $this->searchUserGroups($connectAdmin, $userDn, $userLdapAttributes, $user->login);
        ldap_close($connectAdmin);
        if ($userGroups !== false) {
            // the user is at least in a ldap group, so we synchronize ldap groups
            // with jAcl2 groups
            $this->synchronizeAclGroups($user->login, $userGroups);
        }
        return $user;
    }

    protected function synchronizeAclGroups($login, $userGroups)
    {
        if ($this->_params['searchGroupKeepUserInDefaultGroups']) {
            // Add default groups
            $gplist = jDao::get('jacl2db~jacl2group', 'jacl2_profile')
                ->getDefaultGroups();
            foreach ($gplist as $group) {
                $idx = array_search($group->name, $userGroups);
                if ($idx === false) {
                    $userGroups[] = $group->name;
                }
            }
        }

        // we know the user group: we should be sure it is the same in jAcl2
        $gplist = jDao::get('jacl2db~jacl2groupsofuser', 'jacl2_profile')
            ->getGroupsUser($login);
        $groupsToRemove = array();
        foreach ($gplist as $group) {
            if ($group->grouptype == 2) { // private group
                continue;
            }
            $idx = array_search($group->name, $userGroups);
            if ($idx !== false) {
                unset($userGroups[$idx]);
            } else {
                $groupsToRemove[] = $group->name;
            }
        }
        foreach ($groupsToRemove as $group) {
            jAcl2DbUserGroup::removeUserFromGroup($login, $group);
        }
        foreach ($userGroups as $newGroup) {
            if (jAcl2DbUserGroup::getGroup($newGroup)) {
                jAcl2DbUserGroup::addUserToGroup($login, $newGroup);
            }
        }
    }

    /**
     * @return string[]|false  ldap attributes or false if not found
     */
    protected function searchLdapUserAttributes($connect, $login, $user)
    {
        $searchAttributes = array_keys($this->_params['searchAttributes']);
        $filters = array();
        foreach ($this->_params['searchUserFilter'] as $searchUserFilter) {
            $filter = str_replace(
                array('%%LOGIN%%', '%%USERNAME%%'), // USERNAME deprecated
                $login,
                $searchUserFilter
            );
            $search = @ldap_search(
                $connect,
                $this->_params['searchUserBaseDN'],
                $filter,
                $searchAttributes
            );
            if (!$search) {
                $this->logLdapError($connect, 'ldap error during search of the user "'.$login.'" with the filter "'.$filter.'"');
            }
            else if (($entry = @ldap_first_entry($connect, $search))) {
                $attributes = ldap_get_attributes($connect, $entry);
                return $this->readLdapAttributes($attributes, $user);
            }
            else {
                $filters[] = $filter;
            }
        }
        jLog::log('ldapdao error: ldap user "'.$login.'" not found with the filters :"'.implode('", "', $filters).'"', 'auth');
        return false;
    }

    protected $bindUserDnTries = array();

    protected function bindUser($connect, $userAttributes, $login, $password)
    {
        $bind = false;
        $this->bindUserDnTries = array();
        foreach ($this->_params['bindUserDN'] as $dn) {
            if (preg_match('/^\\$\w+$/', trim($dn))) {
                $dnAttribute = substr($dn, 1);
                if (isset($userAttributes[$dnAttribute])) {
                    $realDn = $userAttributes[$dnAttribute];
                } else {
                    continue;
                }
            } elseif (preg_match_all('/(\w+)=%\?%/', $dn, $m)) {
                $realDn = $dn;
                foreach ($m[1] as $k => $attr) {
                    if (isset($userAttributes[$attr])) {
                        $realDn = str_replace($m[0][$k], $attr.'='.$userAttributes[$attr], $realDn);
                    } else {
                        continue 2;
                    }
                }
            } else {
                $realDn = str_replace(
                    array('%%LOGIN%%', '%%USERNAME%%'), // USERNAME deprecated
                    $login,
                    $dn
                );
            }
            $bind = @ldap_bind($connect, $realDn, $password);
            if ($bind) {
                break;
            } else {
                $this->bindUserDnTries[] = $this->getLdapError($connect, 'tried to connect with "'.$realDn.'"');
            }
        }
        return ($bind ? $realDn : false);
    }

    protected function checkAdminLogin($user, $dao, $password)
    {
        if (!$user) {
            return false;
        }

        $result = $this->checkPassword($password, $user->password);

        if ($result === false) {
            jLog::log('ldapdao: given admin user password is wrong', 'auth');
            return false;
        }

        if ($result !== true) {
            // it is a new hash for the password, let's update it persistently
            $user->password = $result;
            $dao->updatePassword($this->_params['jelixAdminLogin'], $result);
        }
        return $user;
    }

    protected function readLdapAttributes($attributes, $user)
    {
        $mapping = $this->_params['searchAttributes'];
        $ldapAttributes = array();
        foreach ($attributes as $ldapAttr => $attr) {
            if (isset($attr['count']) && $attr['count'] > 0) {
                if ($attr['count'] > 1) {
                    $val = array_shift($attr);
                } else {
                    $val = $attr[0];
                }
                $ldapAttributes[$ldapAttr] = $val;
                if (isset($mapping[$ldapAttr])) {
                    $objAttr = $mapping[$ldapAttr];
                    unset($mapping[$ldapAttr]);
                    if ($objAttr != '') {
                        $user->$objAttr = $val;
                    }
                }
            }
        }

        foreach ($mapping as $ldapAttr => $objAttr) {
            if ($objAttr != '' && !isset($user->$objAttr)) {
                $user->$objAttr = '';
            }
        }
        return $ldapAttributes;
    }

    protected function searchUserGroups($connect, $userDn, $userLdapAttributes, $login)
    {
        // Do not search for groups if no group filter passed
        // Usefull to forbid the driver to sync groups from LDAP and loose all related groups for the user
        if ($this->_params['searchGroupFilter'] == '') {
            return false;
        }

        $searchStr = array_keys($userLdapAttributes);
        $searchStr[] = 'USERDN';
        $searchStr[] = 'LOGIN';
        $searchStr[] = 'USERNAME'; // USERNAME deprecated
        $searchStr = array_map(function ($val) {
            return '%%'.$val.'%%';
        }, $searchStr);
        $values = array_values($userLdapAttributes);
        // escape parenthesis
        $values = array_map(function ($val) {
            return str_replace(
                array('(', ')'),
                array('\\(', '\\)'),
                $val
            );
         }, $values);
        $values[] = $userDn;
        $values[] = $login;
        $values[] = $login;

        $filter = str_replace(
            $searchStr,
            $values,
            $this->_params['searchGroupFilter']
        );
        $grpProp = $this->_params['searchGroupProperty'];

        $groups = array();
        $search = @ldap_search (
            $connect,
            $this->_params['searchGroupBaseDN'],
            $filter,
            array($grpProp)
        );
        if ($search) {
            $entry = @ldap_first_entry($connect, $search);
            if ($entry) {
                do {
                    $attributes = ldap_get_attributes($connect, $entry);
                    if (isset($attributes[$grpProp]) && $attributes[$grpProp]['count'] > 0) {
                        $groups[] = $attributes[$grpProp][0];
                    }
                } while ($entry = ldap_next_entry($connect, $entry));
            }
            else {
                jLog::log('ldapdao: no groups found for the user  "'.$login.'", with the searchGroupFilter query "'.$filter.'"', 'auth');
            }
        }
        else {
            $this->logLdapError($connect, 'ldap error during group search for "'.$login.'", with the searchGroupFilter query "'.$filter.'"');
        }
        return $groups;
    }

    /**
     * open the connection to the ldap server
     */
    protected function _getLinkId()
    {
        if ($connect = @ldap_connect($this->uriConnect)) {
            //ldap_set_option(NULL, LDAP_OPT_DEBUG_LEVEL, 7);
            ldap_set_option($connect, LDAP_OPT_PROTOCOL_VERSION, $this->_params['protocolVersion']);
            ldap_set_option($connect, LDAP_OPT_REFERRALS, 0);

            if ($this->_params['tlsMode'] == 'starttls') {
                if (!@ldap_start_tls($connect)) {
                    $this->logLdapError($connect, 'connection error: impossible to start TLS connection to '.$this->uriConnect);
                    return false;
                }
            }
            return $connect;
        }
        jLog::log('ldapdao error: ldap error, bad syntax in the given uri "'.$this->uriConnect.'"', 'auth');
        return false;
    }

    /**
     * open the connection to the ldap server
     * and bind to the admin user
     * @return resource|false the ldap connection
     */
    protected function _bindLdapAdminUser()
    {
        $connect = $this->_getLinkId();
        if (!$connect) {
            return false;
        }

        if ($this->_params['adminUserDn'] == '') {
            $bind = @ldap_bind($connect);
        } else {
            $bind = @ldap_bind($connect, $this->_params['adminUserDn'], $this->_params['adminPassword']);
        }
        if (!$bind) {
            if ($this->_params['adminUserDn'] == '') {
                jLog::log('ldapdao error: authenticating to the ldap with an anonymous admin user is not supported', 'auth');
            } else {
                jLog::log('ldapdao error: impossible to authenticate to ldap with the admin user '.$this->_params['adminUserDn'], 'auth');
            }
            ldap_close($connect);
            return false;
        }
        return $connect;
    }

    protected function getLdapError($connect, $contextMessage)
    {
        $message = "ldapdao error: $contextMessage \n";
        $message .= "\tldap error " . ldap_errno($connect) . ':' . ldap_error($connect);
        if (@ldap_get_option($connect, LDAP_OPT_DIAGNOSTIC_MESSAGE, $diagnostic)) {
            $message .= "\n\t" . $diagnostic;
        }
        return $message;
    }

    protected function logLdapError($connect, $contextMessage)
    {
        jLog::log($this->getLdapError($connect, $contextMessage), 'auth');
    }
}
